pkg/clientcmd/configutil.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package clientcmd contains convenience methods for working with the kubeconfig and loading specific configurations
+// of api.Config and rest.Config.
+package clientcmd
+
+import (
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+)
+
+// NewApiConfigForDefaultSearchPath creates an api.Config by searching for the kubeconfig on the default search path. If an override 'kubeconfigPath' is
+// given then that path is used instead of the default path.
+func NewApiConfigForDefaultSearchPath(kubeconfigPath string) (*api.Config, error) {
+	configLoader := clientcmd.NewDefaultClientConfigLoadingRules()
+	configLoader.ExplicitPath = kubeconfigPath
+	return configLoader.Load()
+}
+
+// NewApiConfigForKubeconfig creates an api.Config for a given kubeconfig string.
+func NewApiConfigForKubeconfig(kubeconfig string) (*api.Config, error) {
+	return clientcmd.Load([]byte(kubeconfig))
+}
+
+// NewRestConfigForDefaultSearchPath creates a rest.Config by searching for the kubeconfig on the default search path. If an override 'kubeconfigPath' is
+// given then that path is used instead of the default path. If no override is given, an attempt is made to load the
+// 'in cluster' config. If this fails, then the default search path is used.
+func NewRestConfigForDefaultSearchPath(kubeconfigPath string) (*rest.Config, error) {
+	if kubeconfigPath == "" {
+		config, err := rest.InClusterConfig()
+		// if there is no err, continue because InClusterConfig is only expected to succeed if running inside of a pod.
+		if err == nil {
+			return config, nil
+		}
+	}
+	apiConfig, err := NewApiConfigForDefaultSearchPath(kubeconfigPath)
+	if err != nil {
+		return nil, err
+	}
+	return NewRestConfig(apiConfig)
+}
+
+// NewRestConfigForKubeconfig creates a rest.Config for a given kubeconfig string.
+func NewRestConfigForKubeconfig(kubeconfig string) (*rest.Config, error) {
```

Comment:
The reason why they are public, especially the rest.config, is for two reasons:
  1. **Re-using a specific, scoped rest.config**: One use-case that kubectl delete will make use of is selecting a specific cluster from kubeconfig. We will allow the user to supply a cluster name and from that we will select the specific cluster they want to delete from the kubeconfig and not just delete the default / first cluster. It is a bit better to do the work of finding/creating the correct rest.Config for the named cluster once and then re-using that rest.Config to create both the cluster-api and core kubernetes clients. It is a bit easier to reason about when you know that the same rest.Config is used for both. If we reload the rest.Config for both client creations then it is possible for the two clients to use different rest.Configs which creates a bit of complexity. Maybe this is extremely unlikely edge case is not worth worrying about but it is there.
2. Flexibility to create new methods to load a specific cluster's config by IP, cluster name, etc, but I guess in the end those methods should be in this file anwyay so we don't need these to be public. 

But thinking more about it, maybe what we need is methods that return both the cluster-api client AND the core kubernetes client. And that's good enough. For now I will make these private and think some more about this in the "kubecdtl delete" PR. I'll add you to that PR in case I make these public again so you can weigh in if it is appropriate or not.